pkg/client/optimizely_user_context_test.go

```diff
@@ -119,6 +137,574 @@ func (s *OptimizelyUserContextTestSuite) TestSetAttributeOverride() {
+	variationID2 := "10418510624"
+	variationKey1 := "variation_with_traffic"
+	variationKey2 := "variation_no_traffic"
+
+	userProfileService := new(MockUserProfileService)
+	s.OptimizelyClient, _ = s.factory.Client(
+		WithEventProcessor(s.eventProcessor),
+		WithUserProfileService(userProfileService),
+	)
+
+	decisionKey := decision.NewUserDecisionKey(experimentID)
+	savedUserProfile := decision.UserProfile{
+		ID:                  s.userID,
+		ExperimentBucketMap: map[decision.UserDecisionKey]string{decisionKey: variationID2},
+	}
+	userProfileService.On("Lookup", s.userID).Return(savedUserProfile)
+	userProfileService.On("Save", mock.Anything)
+
+	userContext := s.OptimizelyClient.CreateUserContext(s.userID, map[string]interface{}{})
+	decision := userContext.Decide(flagKey, decide.OptimizelyDecideOptions{})
+	// should return variationId2 set by UPS
+	s.Equal(variationKey2, decision.GetVariationKey())
+
+	decision = userContext.Decide(flagKey, decide.OptimizelyDecideOptions{IgnoreUserProfileService: true})
+	// should ignore variationId2 set by UPS and return variationId1
+	s.Equal(variationKey1, decision.GetVariationKey())
+	// also should not save either
+	userProfileService.AssertNotCalled(s.T(), "Save", mock.Anything)
+}
+
+func (s *OptimizelyUserContextTestSuite) TestDecideOptionsExcludeVariables() {
+	flagKey := "feature_1"
+	var options decide.OptimizelyDecideOptions
+	user := s.OptimizelyClient.CreateUserContext(s.userID, nil)
+
+	decision := user.Decide(flagKey, options)
+	s.True(len(decision.GetVariables().ToMap()) > 0)
+
+	options.ExcludeVariables = true
+	decision = user.Decide(flagKey, options)
+	s.Len(decision.GetVariables().ToMap(), 0)
+}
+
+func (s *OptimizelyUserContextTestSuite) TestDecideOptionsIncludeReasons() {
+	flagKey := "invalid_key"
+	var options decide.OptimizelyDecideOptions
+	user := s.OptimizelyClient.CreateUserContext(s.userID, nil)
+
+	// invalid flag key
+	decision := user.Decide(flagKey, options)
+	s.Len(decision.GetReasons(), 1)
+	s.Equal(decide.GetDecideMessage(decide.FlagKeyInvalid, flagKey), decision.GetReasons()[0])
+
+	// invalid flag key with includeReasons
+	options.IncludeReasons = true
+	decision = user.Decide(flagKey, options)
+	s.Len(decision.GetReasons(), 1)
+	s.Equal(decide.GetDecideMessage(decide.FlagKeyInvalid, flagKey), decision.GetReasons()[0])
+
+	// valid flag key
+	options.IncludeReasons = false
+	flagKey = "feature_1"
+	decision = user.Decide(flagKey, options)
+	s.Len(decision.GetReasons(), 0)
+}
+
+func (s *OptimizelyUserContextTestSuite) TestDefaultDecideOptions() {
+	flagKey := "feature_1"
+	options := decide.OptimizelyDecideOptions{ExcludeVariables: true}
+	client, _ := s.factory.Client(WithEventProcessor(s.eventProcessor), WithDefaultDecideOptions(options))
+	userContext := client.CreateUserContext(s.userID, nil)
+
+	// should be excluded by DefaultDecideOption
+	decision := userContext.Decide(flagKey, decide.OptimizelyDecideOptions{})
+	s.Len(decision.GetVariables().ToMap(), 0)
+}
```

Comment:
Can we add one more test with DefaultDecideOptions with "EnabledFlagsOnly". It takes a different path from other options. Worth to have the test.

Reply:
Also need some tests validating DefaultDecideOptions and api parameter options are merged and work together ok.